File: sdk/typescript/src/bridge-values.ts
import { A3SBoxError } from './errors.js'
import type { BridgeResult } from './runtime.js'
import type {
  BuildImageInfo,
  FilesystemSnapshotSummary,
  ImageHealthCheckInfo,
  ImageHistoryInfo,
  ImageInfo,
  ImageInspectInfo,
  NetworkEndpointInfo,
  NetworkInfo,
  PushImageInfo,
  RuntimeDiagnostics,
  RuntimeDiskUsage,
  SandboxLogEntry,
  SandboxStats,
  SandboxSummary,
  SdkCapabilities,
  VolumeInfo,
} from './client.js'
import type {
  EntryInfo,
  FilesystemSnapshotInfo,
} from './sandbox.js'

export function buildImageInfo(result: BridgeResult): BuildImageInfo {
  return {
    reference: requiredString(result, 'reference'),
    digest: requiredString(result, 'digest'),
    sizeBytes: requiredNumber(result, 'size_bytes'),
    layerCount: requiredNumber(result, 'layer_count'),
  }
}

export function imageInfo(result: BridgeResult): ImageInfo {
  return {
    reference: requiredString(result, 'reference'),
    digest: requiredString(result, 'digest'),
    sizeBytes: requiredNumber(result, 'size_bytes'),
    pulledAt: requiredString(result, 'pulled_at'),
    lastUsed: requiredString(result, 'last_used'),
    path: requiredString(result, 'path'),
  }
}

export function imageInspectInfo(result: BridgeResult): ImageInspectInfo {
  const healthCheck =
    result.health_check === null || result.health_check === undefined
      ? undefined
      : imageHealthCheckInfo(asRecord(result.health_check))
  return {
    ...imageInfo(result),
    manifestDigest: requiredString(result, 'manifest_digest'),
    layerCount: requiredNumber(result, 'layer_count'),
    entrypoint: optionalStringArray(result.entrypoint),
    command: optionalStringArray(result.command),
    env: stringRecord(result.env),
    workingDir: optionalString(result, 'working_dir'),
    user: optionalString(result, 'user'),
    exposedPorts: stringArray(result.exposed_ports),
    volumes: stringArray(result.volumes),
    stopSignal: optionalString(result, 'stop_signal'),
    healthCheck,
    onbuild: stringArray(result.onbuild),
    labels: stringRecord(result.labels),
  }
}

export function imageHealthCheckInfo(
  result: BridgeResult
): ImageHealthCheckInfo {
  return {
    test: stringArray(result.test),
    interval: optionalNumber(result, 'interval'),
    timeout: optionalNumber(result, 'timeout'),
    retries: optionalNumber(result, 'retries'),
    startPeriod: optionalNumber(result, 'start_period'),
  }
}

export function imageHistoryInfo(result: BridgeResult): ImageHistoryInfo {
  return {
    created: optionalString(result, 'created'),
    createdBy: requiredString(result, 'created_by'),
    sizeBytes: requiredNumber(result, 'size_bytes'),
    comment: requiredString(result, 'comment'),
    emptyLayer: requiredBoolean(result, 'empty_layer'),
  }
}

export function pushImageInfo(result: BridgeResult): PushImageInfo {
  return {
    reference: requiredString(result, 'reference'),
    manifestDigest: requiredString(result, 'manifest_digest'),
    configUrl: requiredString(result, 'config_url'),
    manifestUrl: requiredString(result, 'manifest_url'),
  }
}

export function sdkCapabilities(result: BridgeResult): SdkCapabilities {
  return {
    protocolVersion: requiredNumber(result, 'protocol_version'),
    operations: stringArray(result.operations),
  }
}

export function volumeInfo(result: BridgeResult): VolumeInfo {
  return {
    name: requiredString(result, 'name'),
    driver: requiredString(result, 'driver'),
    mountPoint: requiredString(result, 'mount_point'),
    labels: stringRecord(result.labels),
    inUseBy: stringArray(result.in_use_by),
    inUse: requiredBoolean(result, 'in_use'),
    sizeLimit: requiredNumber(result, 'size_limit'),
    createdAt: requiredString(result, 'created_at'),
  }
}

export function networkInfo(result: BridgeResult): NetworkInfo {
  return {
    name: requiredString(result, 'name'),
    driver: requiredString(result, 'driver'),
    subnet: requiredString(result, 'subnet'),
    gateway: requiredString(result, 'gateway'),
    labels: stringRecord(result.labels),
    endpoints: recordArray(result, 'endpoints').map(networkEndpointInfo),
    endpointCount: requiredNumber(result, 'endpoint_count'),
    isolation: requiredString(result, 'isolation'),
    createdAt: requiredString(result, 'created_at'),
  }
}

export function networkEndpointInfo(
  result: BridgeResult
): NetworkEndpointInfo {
  return {
    boxId: requiredString(result, 'box_id'),
    boxName: requiredString(result, 'box_name'),
    aliases: stringArray(result.aliases),
    ipAddress: requiredString(result, 'ip_address'),
    macAddress: requiredString(result, 'mac_address'),
  }
}

export function sandboxSummary(result: BridgeResult): SandboxSummary {
  return {
    id: requiredString(result, 'id'),
    shortId: requiredString(result, 'short_id'),
    name: requiredString(result, 'name'),
    image: requiredString(result, 'image'),
    isolation: requiredString(result, 'isolation'),
    status: requiredString(result, 'status'),
    statusSummary: requiredString(result, 'status_summary'),
    active: requiredBoolean(result, 'active'),
    pid: optionalNumber(result, 'pid'),
    cpus: requiredNumber(result, 'cpus'),
    memoryMb: requiredNumber(result, 'memory_mb'),
    ports: stringArray(result.ports),
    command: stringArray(result.command),
    health: requiredString(result, 'health'),
    labels: stringRecord(result.labels),
    createdAt: requiredString(result, 'created_at'),
    startedAt: optionalString(result, 'started_at'),
    networkName: optionalString(result, 'network_name'),
    volumeNames: stringArray(result.volume_names),
  }
}

export function sandboxLogEntry(result: BridgeResult): SandboxLogEntry {
  return {
    stream: requiredString(result, 'stream'),
    message: requiredString(result, 'log'),
    timestamp: optionalString(result, 'time'),
  }
}

export function sandboxStats(result: BridgeResult): SandboxStats {
  return {
    id: requiredString(result, 'id'),
    shortId: requiredString(result, 'short_id'),
    name: requiredString(result, 'name'),
    status: requiredString(result, 'status'),
    pid: requiredNumber(result, 'pid'),
    cpus: requiredNumber(result, 'cpus'),
    cpuPercent: requiredNumber(result, 'cpu_percent'),
    cpuPercentScaled: requiredNumber(result, 'cpu_percent_scaled'),
    memoryBytes: requiredNumber(result, 'memory_bytes'),
    memoryLimitBytes: requiredNumber(result, 'memory_limit_bytes'),
    memoryPercent: requiredNumber(result, 'memory_percent'),
    networkRxBytes: requiredNumber(result, 'network_rx_bytes'),
    networkTxBytes: requiredNumber(result, 'network_tx_bytes'),
    blockReadBytes: requiredNumber(result, 'block_read_bytes'),
    blockWriteBytes: requiredNumber(result, 'block_write_bytes'),
  }
}

export function runtimeDiagnostics(result: BridgeResult): RuntimeDiagnostics {
  const virtualization = asRecord(result.virtualization)
  return {
    coreVersion: requiredString(result, 'core_version'),
    runtimeVersion: requiredString(result, 'runtime_version'),
    sdkVersion: requiredString(result, 'sdk_version'),
    home: requiredString(result, 'home'),
    virtualization: {
      available: requiredBoolean(virtualization, 'available'),
      backend: optionalString(virtualization, 'backend'),
      details: requiredString(virtualization, 'details'),
    },
  }
}

export function runtimeDiskUsage(result: BridgeResult): RuntimeDiskUsage {
  return {
    home: requiredString(result, 'home'),
    totalBytes: requiredNumber(result, 'total_bytes'),
    boxesBytes: requiredNumber(result, 'boxes_bytes'),
    imagesBytes: requiredNumber(result, 'images_bytes'),
    volumesBytes: requiredNumber(result, 'volumes_bytes'),
    snapshotsBytes: requiredNumber(result, 'snapshots_bytes'),
    stateBytes: requiredNumber(result, 'state_bytes'),
    otherBytes: requiredNumber(result, 'other_bytes'),
  }
}

export function filesystemSnapshotSummary(
  result: BridgeResult
): FilesystemSnapshotSummary {
  return {
    id: requiredString(result, 'id'),
    name: requiredString(result, 'name'),
    sourceSandboxId: requiredString(result, 'source_box_id'),
    image: requiredString(result, 'image'),
    vcpus: requiredNumber(result, 'vcpus'),
    memoryMb: requiredNumber(result, 'memory_mb'),
    volumes: stringArray(result.volumes),
    command: stringArray(result.command),
    ports: stringArray(result.port_map),
    labels: stringRecord(result.labels),
    networkMode: optionalString(result, 'network_mode'),
    sizeBytes: requiredNumber(result, 'size_bytes'),
    createdAt: requiredString(result, 'created_at'),
    description: requiredString(result, 'description'),
  }
}

export function entryInfo(entry: BridgeResult): EntryInfo {
  return {
    name: requiredString(entry, 'name'),
    type: entryType(entry.type),
    path: requiredString(entry, 'path'),
    size: requiredNumber(entry, 'size'),
    mode: requiredNumber(entry, 'mode'),
    permissions: requiredString(entry, 'permissions'),
    owner: requiredString(entry, 'owner'),
    group: requiredString(entry, 'group'),
    modifiedSeconds: requiredNumber(entry, 'modified_seconds'),
    modifiedNanos: requiredNumber(entry, 'modified_nanos'),
    ...(typeof entry.symlink_target === 'string'
      ? { symlinkTarget: entry.symlink_target }
      : {}),
  }
}

export function filesystemSnapshotInfo(
  result: BridgeResult
): FilesystemSnapshotInfo {
  return {
    snapshotId: requiredString(result, 'snapshot_id'),
    sizeBytes: requiredNumber(result, 'size_bytes'),
    state: requiredString(result, 'state'),
    generation: requiredNumber(result, 'generation'),
  }
}

function entryType(value: unknown): EntryInfo['type'] {
  if (value === 'file' || value === 'directory' || value === 'unspecified') {
    return value
  }
  throw new A3SBoxError(
    'Bridge returned an invalid entry type',
    'bridge_protocol_error'
  )
}

export function requiredString(
  result: BridgeResult,
  key: string
): string {
  const value = result[key]
  if (typeof value !== 'string') bridgeTypeError(key)
  return value
}

export function requiredNumber(
  result: BridgeResult,
  key: string
): number {
  const value = result[key]
  if (typeof value !== 'number') bridgeTypeError(key)
  return value
}

export function requiredBoolean(
  result: BridgeResult,
  key: string
): boolean {
  const value = result[key]
  if (typeof value !== 'boolean') bridgeTypeError(key)
  return value
}

export function optionalString(
  result: BridgeResult,
  key: string
): string | undefined {
  const value = result[key]
  if (value === null || value === undefined) return undefined
  if (typeof value !== 'string') bridgeTypeError(key)
  return value
}

export function optionalNumber(
  result: BridgeResult,
  key: string
): number | undefined {
  const value = result[key]
  if (value === null || value === undefined) return undefined
  if (typeof value !== 'number') bridgeTypeError(key)
  return value
}

export function recordArray(
  result: BridgeResult,
  key: string
): BridgeResult[] {
  const value = result[key]
  if (!Array.isArray(value)) bridgeTypeError(key)
  return value.map(asRecord)
}

export function unknownRecordArray(value: unknown): BridgeResult[] {
  if (!Array.isArray(value)) bridgeTypeError('array')
  return value.map(asRecord)
}

export function stringArray(value: unknown): string[] {
  if (!Array.isArray(value) || !value.every((item) => typeof item === 'string')) {
    bridgeTypeError('array')
  }
  return value
}

export function optionalStringArray(
  value: unknown
): string[] | undefined {
  return value === null || value === undefined ? undefined : stringArray(value)
}

export function stringRecord(
  value: unknown
): Readonly<Record<string, string>> {
  const record = asRecord(value)
  if (!Object.values(record).every((item) => typeof item === 'string')) {
    bridgeTypeError('record')
  }
  return record as Record<string, string>
}

export function requiredRecord(
  result: BridgeResult,
  key: string
): BridgeResult {
  return asRecord(result[key])
}

export function asRecord(value: unknown): BridgeResult {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    bridgeTypeError('object')
  }
  return value as BridgeResult
}

export function decodeBase64(result: BridgeResult, key: string): Buffer {
  return Buffer.from(requiredString(result, key), 'base64')
}

function bridgeTypeError(key: string): never {
  throw new A3SBoxError(
    `Bridge result has an invalid ${key}`,
    'bridge_protocol_error'
  )
}

File: sdk/python/src/a3s_box/_bridge_values.py
"""Strict decoders for values returned by the local machine bridge."""

from __future__ import annotations

import base64
import binascii
from collections.abc import Mapping, Sequence
from typing import Literal, NoReturn, cast

from .exceptions import A3SBoxError
from .models import (
    BuildImageInfo,
    CommandResult,
    EntryInfo,
    ExecutionCpuStats,
    ExecutionEventBatch,
    ExecutionEventKind,
    ExecutionMemoryStats,
    ExecutionProcessInfo,
    ExecutionProcessInventory,
    ExecutionRuntimeEvent,
    ExecutionStats,
    FilesystemSnapshotInfo,
    FilesystemSnapshotSummary,
    ImageHealthCheckInfo,
    ImageHistoryInfo,
    ImageInfo,
    ImageInspectInfo,
    NetworkEndpointInfo,
    NetworkInfo,
    PushImageInfo,
    RuntimeDiagnostics,
    RuntimeDiskUsage,
    RuntimeVirtualization,
    SandboxLogEntry,
    SandboxStats,
    SandboxSummary,
    SdkCapabilities,
    VolumeInfo,
)


def build_image_info(result: Mapping[str, object]) -> BuildImageInfo:
    return BuildImageInfo(
        reference=string(result["reference"]),
        digest=string(result["digest"]),
        size_bytes=integer(result["size_bytes"]),
        layer_count=integer(result["layer_count"]),
    )


def image_info(result: Mapping[str, object]) -> ImageInfo:
    return ImageInfo(
        reference=string(result["reference"]),
        digest=string(result["digest"]),
        size_bytes=integer(result["size_bytes"]),
        pulled_at=string(result["pulled_at"]),
        last_used=string(result["last_used"]),
        path=string(result["path"]),
    )


def image_inspect_info(result: Mapping[str, object]) -> ImageInspectInfo:
    health_value = result.get("health_check")
    return ImageInspectInfo(
        reference=string(result["reference"]),
        digest=string(result["digest"]),
        size_bytes=integer(result["size_bytes"]),
        pulled_at=string(result["pulled_at"]),
        last_used=string(result["last_used"]),
        path=string(result["path"]),
        manifest_digest=string(result["manifest_digest"]),
        layer_count=integer(result["layer_count"]),
        entrypoint=optional_string_tuple(result.get("entrypoint")),
        command=optional_string_tuple(result.get("command")),
        env=string_mapping(result["env"]),
        working_dir=optional_string(result.get("working_dir")),
        user=optional_string(result.get("user")),
        exposed_ports=string_tuple(result["exposed_ports"]),
        volumes=string_tuple(result["volumes"]),
        stop_signal=optional_string(result.get("stop_signal")),
        health_check=(
            None
            if health_value is None
            else image_health_check_info(mapping(health_value))
        ),
        onbuild=string_tuple(result["onbuild"]),
        labels=string_mapping(result["labels"]),
    )


def image_health_check_info(
    result: Mapping[str, object],
) -> ImageHealthCheckInfo:
    return ImageHealthCheckInfo(
        test=string_tuple(result["test"]),
        interval=optional_int(result.get("interval")),
        timeout=optional_int(result.get("timeout")),
        retries=optional_int(result.get("retries")),
        start_period=optional_int(result.get("start_period")),
    )


def image_history_info(result: Mapping[str, object]) -> ImageHistoryInfo:
    return ImageHistoryInfo(
        created=optional_string(result.get("created")),
        created_by=string(result["created_by"]),
        size_bytes=integer(result["size_bytes"]),
        comment=string(result["comment"]),
        empty_layer=boolean(result["empty_layer"]),
    )


def push_image_info(result: Mapping[str, object]) -> PushImageInfo:
    return PushImageInfo(
        reference=string(result["reference"]),
        manifest_digest=string(result["manifest_digest"]),
        config_url=string(result["config_url"]),
        manifest_url=string(result["manifest_url"]),
    )


def sdk_capabilities(result: Mapping[str, object]) -> SdkCapabilities:
    return SdkCapabilities(
        protocol_version=integer(result["protocol_version"]),
        operations=string_tuple(result["operations"]),
    )


def volume_info(result: Mapping[str, object]) -> VolumeInfo:
    return VolumeInfo(
        name=string(result["name"]),
        driver=string(result["driver"]),
        mount_point=string(result["mount_point"]),
        labels=string_mapping(result["labels"]),
        in_use_by=string_tuple(result["in_use_by"]),
        in_use=boolean(result["in_use"]),
        size_limit=integer(result["size_limit"]),
        created_at=string(result["created_at"]),
    )


def network_info(result: Mapping[str, object]) -> NetworkInfo:
    return NetworkInfo(
        name=string(result["name"]),
        driver=string(result["driver"]),
        subnet=string(result["subnet"]),
        gateway=string(result["gateway"]),
        labels=string_mapping(result["labels"]),
        endpoints=tuple(
            network_endpoint(item)
            for item in mapping_sequence(result["endpoints"])
        ),
        endpoint_count=integer(result["endpoint_count"]),
        isolation=string(result["isolation"]),
        created_at=string(result["created_at"]),
    )


def network_endpoint(result: Mapping[str, object]) -> NetworkEndpointInfo:
    return NetworkEndpointInfo(
        box_id=string(result["box_id"]),
        box_name=string(result["box_name"]),
        aliases=string_tuple(result["aliases"]),
        ip_address=string(result["ip_address"]),
        mac_address=string(result["mac_address"]),
    )


def sandbox_summary(result: Mapping[str, object]) -> SandboxSummary:
    return SandboxSummary(
        id=string(result["id"]),
        short_id=string(result["short_id"]),
        name=string(result["name"]),
        image=string(result["image"]),
        isolation=string(result["isolation"]),
        status=string(result["status"]),
        status_summary=string(result["status_summary"]),
        active=boolean(result["active"]),
        pid=optional_int(result.get("pid")),
        cpus=integer(result["cpus"]),
        memory_mb=integer(result["memory_mb"]),
        ports=string_tuple(result["ports"]),
        command=string_tuple(result["command"]),
        health=string(result["health"]),
        labels=string_mapping(result["labels"]),
        created_at=string(result["created_at"]),
        started_at=optional_string(result.get("started_at")),
        network_name=optional_string(result.get("network_name")),
        volume_names=string_tuple(result["volume_names"]),
    )


def sandbox_log_entry(result: Mapping[str, object]) -> SandboxLogEntry:
    return SandboxLogEntry(
        stream=string(result["stream"]),
        message=string(result["log"]),
        timestamp=optional_string(result.get("time")),
    )


def sandbox_stats(result: Mapping[str, object]) -> SandboxStats:
    return SandboxStats(
        id=string(result["id"]),
        short_id=string(result["short_id"]),
        name=string(result["name"]),
        status=string(result["status"]),
        pid=integer(result["pid"]),
        cpus=integer(result["cpus"]),
        cpu_percent=number(result["cpu_percent"]),
        cpu_percent_scaled=number(result["cpu_percent_scaled"]),
        memory_bytes=integer(result["memory_bytes"]),
        memory_limit_bytes=integer(result["memory_limit_bytes"]),
        memory_percent=number(result["memory_percent"]),
        network_rx_bytes=integer(result["network_rx_bytes"]),
        network_tx_bytes=integer(result["network_tx_bytes"]),
        block_read_bytes=integer(result["block_read_bytes"]),
        block_write_bytes=integer(result["block_write_bytes"]),
    )


def execution_process_inventory(
    result: Mapping[str, object],
) -> ExecutionProcessInventory:
    return ExecutionProcessInventory(
        execution_id=string(result["execution_id"]),
        generation=integer(result["generation"]),
        processes=tuple(
            ExecutionProcessInfo(
                process_id=string(item["process_id"]),
                pid=optional_unsigned_integer(item.get("pid")),
                terminal=boolean(item["terminal"]),
            )
            for item in mapping_sequence(result["processes"])
        ),
    )


def execution_stats(result: Mapping[str, object]) -> ExecutionStats:
    cpu = mapping(result["cpu"])
    memory = mapping(result["memory"])
    return ExecutionStats(
        execution_id=string(result["execution_id"]),
        generation=integer(result["generation"]),
        timestamp_unix_ns=unsigned_decimal(result["timestamp_unix_ns"]),
        cpu=ExecutionCpuStats(
            usage_ns=unsigned_integer(cpu["usage_ns"]),
            user_ns=unsigned_integer(cpu["user_ns"]),
            system_ns=unsigned_integer(cpu["system_ns"]),
            throttled_ns=unsigned_integer(cpu["throttled_ns"]),
        ),
        memory=ExecutionMemoryStats(
            usage_bytes=unsigned_integer(memory["usage_bytes"]),
            limit_bytes=optional_unsigned_integer(memory.get("limit_bytes")),
            peak_bytes=optional_unsigned_integer(memory.get("peak_bytes")),
        ),
        process_count=unsigned_integer(result["process_count"]),
        metrics=integer_mapping(result["metrics"]),
    )


def execution_event_batch(
    result: Mapping[str, object],
) -> ExecutionEventBatch:
    return ExecutionEventBatch(
        execution_id=string(result["execution_id"]),
        generation=integer(result["generation"]),
        events=tuple(
            ExecutionRuntimeEvent(
                sequence=unsigned_integer(item["sequence"]),
                timestamp_unix_ns=unsigned_decimal(item["timestamp_unix_ns"]),
                process_id=optional_string(item.get("process_id")),
                kind=execution_event_kind(item["kind"]),
                attributes=string_mapping(item["attributes"]),
            )
            for item in mapping_sequence(result["events"])
        ),
        next_sequence=unsigned_integer(result["next_sequence"]),
    )


def execution_event_kind(value: object) -> ExecutionEventKind:
    kind = string(value)
    if kind not in {
        "container-creating",
        "container-created",
        "container-started",
        "container-stopped",
        "container-deleted",
        "container-paused",
        "container-resumed",
        "resources-updated",
        "process-created",
        "process-started",
        "process-exited",
        "output-dropped",
        "runtime-warning",
    }:
        protocol_error("an invalid execution event kind")
    return cast(ExecutionEventKind, kind)


def runtime_diagnostics(result: Mapping[str, object]) -> RuntimeDiagnostics:
    virtualization = mapping(result["virtualization"])
    return RuntimeDiagnostics(
        core_version=string(result["core_version"]),
        runtime_version=string(result["runtime_version"]),
        sdk_version=string(result["sdk_version"]),
        home=string(result["home"]),
        virtualization=RuntimeVirtualization(
            available=boolean(virtualization["available"]),
            backend=optional_string(virtualization.get("backend")),
            details=string(virtualization["details"]),
        ),
    )


def runtime_disk_usage(result: Mapping[str, object]) -> RuntimeDiskUsage:
    return RuntimeDiskUsage(
        home=string(result["home"]),
        total_bytes=integer(result["total_bytes"]),
        boxes_bytes=integer(result["boxes_bytes"]),
        images_bytes=integer(result["images_bytes"]),
        volumes_bytes=integer(result["volumes_bytes"]),
        snapshots_bytes=integer(result["snapshots_bytes"]),
        state_bytes=integer(result["state_bytes"]),
        other_bytes=integer(result["other_bytes"]),
    )


def filesystem_snapshot_summary(
    result: Mapping[str, object],
) -> FilesystemSnapshotSummary:
    return FilesystemSnapshotSummary(
        id=string(result["id"]),
        name=string(result["name"]),
        source_sandbox_id=string(result["source_box_id"]),
        image=string(result["image"]),
        vcpus=integer(result["vcpus"]),
        memory_mb=integer(result["memory_mb"]),
        volumes=string_tuple(result["volumes"]),
        command=string_tuple(result["command"]),
        ports=string_tuple(result["port_map"]),
        labels=string_mapping(result["labels"]),
        network_mode=optional_string(result.get("network_mode")),
        size_bytes=integer(result["size_bytes"]),
        created_at=string(result["created_at"]),
        description=string(result["description"]),
    )


def filesystem_snapshot_info(
    result: Mapping[str, object],
) -> FilesystemSnapshotInfo:
    return FilesystemSnapshotInfo(
        snapshot_id=string(result["snapshot_id"]),
        size_bytes=integer(result["size_bytes"]),
        state=string(result["state"]),
        generation=integer(result["generation"]),
    )


def command_result(result: Mapping[str, object]) -> CommandResult:
    stdout = decoded_base64(result.get("stdout_base64", ""), "stdout_base64")
    stderr = decoded_base64(result.get("stderr_base64", ""), "stderr_base64")
    return CommandResult(
        stdout=stdout.decode(errors="replace"),
        stderr=stderr.decode(errors="replace"),
        exit_code=integer(result["exit_code"]),
        truncated=boolean(result.get("truncated", False)),
    )


def entry_info(entry: Mapping[str, object]) -> EntryInfo:
    entry_type = string(entry["type"])
    if entry_type not in {"file", "directory", "unspecified"}:
        protocol_error("an invalid entry type")
    return EntryInfo(
        name=string(entry["name"]),
        type=cast(
            Literal["file", "directory", "unspecified"],
            entry_type,
        ),
        path=string(entry["path"]),
        size=integer(entry["size"]),
        mode=integer(entry["mode"]),
        permissions=string(entry["permissions"]),
        owner=string(entry["owner"]),
        group=string(entry["group"]),
        modified_seconds=integer(entry["modified_seconds"]),
        modified_nanos=integer(entry["modified_nanos"]),
        symlink_target=optional_string(entry.get("symlink_target")),
    )


def mapping(value: object) -> Mapping[str, object]:
    if not isinstance(value, Mapping):
        protocol_error("a non-object value")
    return cast(Mapping[str, object], value)


def sequence(value: object) -> Sequence[object]:
    if not isinstance(value, Sequence) or isinstance(
        value,
        (str, bytes, bytearray),
    ):
        protocol_error("a non-array value")
    return cast(Sequence[object], value)


def mapping_sequence(value: object) -> list[Mapping[str, object]]:
    return [mapping(item) for item in sequence(value)]


def mapping_list(
    result: Mapping[str, object],
    key: str,
) -> list[Mapping[str, object]]:
    return mapping_sequence(result[key])


def string_mapping(value: object) -> dict[str, str]:
    result = mapping(value)
    if any(
        not isinstance(key, str) or not isinstance(item, str)
        for key, item in result.items()
    ):
        protocol_error("a non-string mapping")
    return dict(cast(Mapping[str, str], result))


def integer_mapping(value: object) -> dict[str, int]:
    result = mapping(value)
    if any(
        not isinstance(key, str)
        or isinstance(item, bool)
        or not isinstance(item, int)
        or item < 0
        for key, item in result.items()
    ):
        protocol_error("a non-integer mapping")
    return dict(cast(Mapping[str, int], result))


def string(value: object) -> str:
    if not isinstance(value, str):
        protocol_error("a non-string value")
    return value


def string_tuple(value: object) -> tuple[str, ...]:
    values = sequence(value)
    if any(not isinstance(item, str) for item in values):
        protocol_error("a non-string array")
    return tuple(cast(Sequence[str], values))


def optional_string(value: object) -> str | None:
    return None if value is None else string(value)


def optional_string_tuple(value: object) -> tuple[str, ...] | None:
    return None if value is None else string_tuple(value)


def integer(value: object) -> int:
    if isinstance(value, bool) or not isinstance(value, int):
        protocol_error("a non-integer value")
    return value


def unsigned_integer(value: object) -> int:
    result = integer(value)
    if result < 0 or result > (1 << 64) - 1:
        protocol_error("an out-of-range unsigned integer")
    return result


def unsigned_decimal(value: object) -> int:
    encoded = string(value)
    if (
        not encoded
        or len(encoded) > 20
        or (len(encoded) > 1 and encoded.startswith("0"))
        or any(character not in "0123456789" for character in encoded)
    ):
        protocol_error("an invalid unsigned decimal string")
    result = int(encoded)
    if result > (1 << 64) - 1:
        protocol_error("an out-of-range unsigned decimal string")
    return result


def optional_unsigned_integer(value: object) -> int | None:
    return None if value is None else unsigned_integer(value)


def optional_int(value: object) -> int | None:
    return None if value is None else integer(value)


def number(value: object) -> float:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        protocol_error("a non-number value")
    return float(value)


def boolean(value: object) -> bool:
    if not isinstance(value, bool):
        protocol_error("a non-boolean value")
    return value


def decoded_base64(value: object, field: str) -> bytes:
    encoded = string(value)
    try:
        return base64.b64decode(encoded, validate=True)
    except (binascii.Error, ValueError) as error:
        raise A3SBoxError(
            f"A3S Box bridge returned invalid {field}",
            code="bridge_protocol_error",
        ) from error


def protocol_error(detail: str) -> NoReturn:
    raise A3SBoxError(
        f"A3S Box bridge returned {detail}",
        code="bridge_protocol_error",
    )
